lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs

```diff
@@ -0,0 +1,120 @@
+    , Scheme (..)
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.Jormungandr.Api
+    ( BlockId, GetTipId, api )
+import Cardano.Wallet.Network
+    ( ErrNetworkUnreachable (..) )
+import Control.Exception
+    ( Exception )
+import Control.Monad.Catch
+    ( throwM )
+import Control.Monad.Trans.Except
+    ( ExceptT (..) )
+import Data.Proxy
+    ( Proxy (..) )
+import Network.HTTP.Client
+    ( Manager, defaultManagerSettings, newManager )
+import Servant.Client
+    ( BaseUrl (..), ClientM, Scheme (..), client, mkClientEnv, runClientM )
+import Servant.Client.Core
+    ( ServantError (..) )
+import Servant.Links
+    ( Link, safeLink )
+
+-- TODO: Implement a NetworkLayer
+-- -- | Constructs a network layer with the given @Jormungandr@ client.
+-- mkNetworkLayer :: Monad m => Jormungandr m -> NetworkLayer m
+-- mkNetworkLayer _httpBridge = NetworkLayer
+--     { nextBlocks = undefined
+--     , networkTip = undefined
+--     , postTx = undefined
+--     }
+--
+-- -- | Creates a jormungandr 'NetworkLayer' using the given connection
+-- -- settings.
+-- newNetworkLayer
+--     :: Int -> IO (NetworkLayer IO)
+-- newNetworkLayer port = mkNetworkLayer <$> newHttpBridge port
+--   where
+--     newHttpBridge = undefined
+
+{-------------------------------------------------------------------------------
+                            Jormungandr Client
+-------------------------------------------------------------------------------}
+
+-- | Endpoints of the jormungandr REST API.
+newtype Jormungandr m = Jormungandr
+    { getTipId
+        :: ExceptT ErrNetworkUnreachable m BlockId
+    }
+
+-- | Construct a 'Jormungandr'-client
+--
+-- >>> mgr <- newManager defaultManagerSettings
+-- >>> j = mkJormungandr mgr (BaseUrl Http "localhost" 8080 "")
+-- >>> runExceptT $ getTipId j
+-- Right (BlockId (Hash {getHash = "26c640a3de09b74398c14ca0a137ec78"}))
+mkJormungandr
+    :: Manager -> BaseUrl -> Jormungandr IO
+mkJormungandr mgr baseUrl = Jormungandr
+    { getTipId = ExceptT $ do
+        let ctx = safeLink api (Proxy @GetTipId)
+        run cGetTipId >>= defaultHandler ctx
+    }
+  where
+    run :: ClientM a -> IO (Either ServantError a)
+    run query = runClientM query (mkClientEnv mgr baseUrl)
+
+    defaultHandler
+        :: Link
+        -> Either ServantError a
+        -> IO (Either ErrNetworkUnreachable a)
+    defaultHandler ctx = \case
+        Right c -> return $ Right c
+
+        -- The node has not started yet or has exited.
+        -- This could be recovered from by either waiting for the node
+        -- initialise, or restarting the node.
+        Left (ConnectionError e) ->
+            return $ Left $ ErrNetworkUnreachable e
+
+        -- Other errors (status code, decode failure, invalid content type
+        -- headers). These are considered to be programming errors, so crash.
+        Left e -> do
+            throwM (ErrUnexpectedNetworkFailure ctx e)
+
+    cGetTipId = client api
+
+data ErrUnexpectedNetworkFailure
```

Comment:
This is now duplicated.  https://github.com/input-output-hk/cardano-wallet/blob/80bf249a03b40811c818094df8ad42b00041fd89/lib/http-bridge/src/Cardano/Wallet/HttpBridge/Network.hs#L181-L186

We _can_ share it by putting it in `Cardano.Wallet.Network`, if we add `servant-client` as a dependancy to core.

Reply:
Hmmm. I wouldn't. This has nothing to do with the core code and is really just an implementation details of the bridge or jormungandr. Duplication is okay here as it's not the same piece of logic. It's simply two pieces of code that have a similar implementation. It'll be easier to maintain by **not** introducing a wrong abstraction to unify them.